File: frontend/src/lib/state.ts
import { get, writable, type Writable } from "svelte/store";

import { EventsOn } from "../../wailsjs/runtime";

export enum Step { // Each step matches a view in Wally
  PROBING,
  KEYBOARD_SELECT,
  FIRMWARE_SELECT,
  KEYBOARD_RESET,
  FIRMWARE_FLASHING,
  FLASH_COMPLETE,
  FATAL_ERROR,
  WALLY_UPDATE,
  WALLY_UPDATE_COMPLETE,
}

export enum FirmwareFormat {
  HEX,
  BIN,
}

type TProgress = {
  current: number;
  total: number;
};

type Device = {
  bootloader: boolean; // is the device currently in bootloader mode.
  fingerprint: number; // unique id equals to the libusb handle pointer casted to an int
  firmwareFormat: number;
  friendlyName: string; // Friendly name ex: Moonlander MK1
  model: string; // a model identifier used to identify the keyboard ex: moonlander, ergodox, planck, stm32, halfkay, gd32
};

export type TLog = {
  timestamp: number;
  level: "info" | "warning" | "fatal";
  message: string;
};

type TState = {
  appVersion: string;
  checkUpdates: boolean;
  devices: Map<number, Device>;
  flashProgress: number;
  hasError: boolean;
  logs: TLog[];
  selectedDevice: Device | null;
  showAbout: boolean;
  showConsole: boolean;
  step: Step;
  updateProgress: number;
};

const state = writable<TState>({
  appVersion: "",
  checkUpdates: false,
  devices: new Map(),
  flashProgress: 0,
  hasError: false,
  logs: [],
  selectedDevice: null,
  showAbout: false,
  showConsole: false,
  step: Step.PROBING,
  updateProgress: 0,
});

//Attaches the events from the Go process to the UI state
function attachToEvents(state: Writable<TState>) {
  EventsOn("log", ({ log }: { log: TLog }) => {
    state.update((_state: TState) => {
      _state.logs.push(log);
      if (log.level == "fatal") {
        _state.hasError = true;
      }
      return _state;
    });
  });

  EventsOn("reset", () => {
    state.update((_state: TState) => {
      _state.hasError = false;
      _state.selectedDevice = null;
      _state.flashProgress = 0;
      _state.updateProgress = 0;
      return _state;
    });
  });

  EventsOn("stepChanged", ({ step }: { step: Step }) => {
    state.update((_state: TState) => {
      _state.step = step;
      return _state;
    });
  });

  EventsOn("deviceConnected", ({ device }: { device: Device }) => {
    state.update((_state) => {
      _state.devices.set(device.fingerprint, device);
      return _state;
    });
  });

  EventsOn("deviceSelected", ({ device }: { device: Device }) => {
    console.log(device);
    state.update((_state) => {
      _state.selectedDevice = device;
      return _state;
    });
  });

  EventsOn("updateProgress", (progress: TProgress) => {
    state.update((_state) => {
      if (progress.total > 0) {
        _state.updateProgress = Math.floor(
          (progress.current / progress.total) * 100
        );
      }
      return _state;
    });
  });

  EventsOn("flashProgress", (progress: TProgress) => {
    state.update((_state) => {
      if (progress.total > 0) {
        _state.flashProgress = Math.floor(
          (progress.current / progress.total) * 100
        );
      }
      return _state;
    });
  });

  EventsOn("deviceDisconnected", ({ fingerprint }: { fingerprint: number }) => {
    state.update((_state) => {
      _state.devices.delete(fingerprint);
      // If there's no devices connected after removing it from the map
      // Redirect to probing view if the current step is Keyboard select or Firmware select

      if (
        _state.devices.size == 0 &&
        (_state.step == Step.KEYBOARD_SELECT ||
          _state.step == Step.FIRMWARE_SELECT)
      )
        _state.step = Step.PROBING;
      return _state;
    });
  });

  EventsOn("promptUpdateCheck", () => {
    const update = window.confirm(
      "Would you like Wally to check for updates on startup?"
    );
  });
}

attachToEvents(state);

export default state;

File: state/state.go
package state

import (
	"context"
	"fmt"
	"os/exec"
	"runtime"
	"time"

	wails "github.com/wailsapp/wails/v2/pkg/runtime"
	"github.com/zsa/wally/usb"
)

type Step int8

const (
	Probing          Step = iota // probing keyboard
	KeybordSelect                // select keyboard
	FirmwareSelect               // select firmware file
	KeyboardReset                // waiting for keyboard reset
	FirmwareFlashing             // flashing
	FlashComplete                // complete
	FatalError
	WallyUpdate
	WallyUpdateComplete
)

type State struct {
	Devices        map[int]usb.USBDevice `json:"devices"`
	FirmwarePath   *string
	Logs           []log          `json:"logs"`
	SelectedDevice *usb.USBDevice `json:"selectedDevice"`
	Step           Step
	version        string
	ctx            context.Context
	enumerator     usb.Enumerator
	config         Configuration
	updatePath     string
}

func (s *State) InitUSB() {

	InitUIEventEmitter(s.ctx)
	enumerator := usb.NewEnumerator()
	cb := usb.NewDirectorEventHandler(s)
	enumerator.SetEventObject(cb)
	s.enumerator = enumerator

	go func() {
		s.Log("info", "UI started, listening to usb events")
		s.enumerator.ListenDevices()
	}()
	/*
		go func() {
			s.enumerator.HandleEvents()
		}()
	*/
}

func (s *State) GetAppVersion() string {
	return s.version
}

func (s *State) Init(ctx context.Context) {
	s.ctx = ctx
	s.version = GetAppVersion()
	s.config = NewConfiguration()
}

func (s *State) Log(level string, message string) {
	now := time.Now()
	l := log{Timestamp: now.Unix(), Level: level, Message: message}
	s.Logs = append(s.Logs, l)
	uiEvent.Emit("log", &LogEvent{Log: l})
}

func (s *State) SetStep(step Step) {
	s.Step = step
	uiEvent.Emit("stepChanged", &StepChangeEvent{Step: step})
}

func (s *State) SetUpdateCheck(val bool) {
	s.config.SetUpdateCheck(val)
}

func (s *State) GetUpdateCheck() bool {
	return s.config.UpdateCheck
}

func (s *State) SelectDevice(fingerprint int) {
	if device, ok := s.Devices[fingerprint]; ok {
		s.SelectedDevice = &device
		uiEvent.Emit("deviceSelected", &DeviceSelectedEvent{Device: *s.SelectedDevice})
	}
}

func (s *State) SelectFirmware() {

	if s.SelectedDevice == nil {
		//LOG err
		return
	}

	pattern := ""

	if s.SelectedDevice.FirmwareFormat == usb.DeviceBIN {
		pattern = "*.bin"
	}

	if s.SelectedDevice.FirmwareFormat == usb.DeviceHEX {
		pattern = "*.hex"
	}

	if pattern == "" {
		//Log err
		return
	}

	selection, err := wails.OpenFileDialog(s.ctx, wails.OpenDialogOptions{
		Title: "Select firmware file",
		Filters: []wails.FileFilter{{
			DisplayName: "Firmware file (" + pattern + ")",
			Pattern:     pattern,
		}},
	})

	if err != nil {
		//Log err
		return
	}

	if selection != "" {
		s.FirmwarePath = &selection
		// If the selected device is a bootloader start the flashing process
		// else jump to the Keyboard Reset screen
		if s.SelectedDevice.Bootloader {
			s.StartFlashing()
		} else {
			s.SetStep(KeyboardReset)
		}
	}
}

func (s *State) StartFlashing() {
	s.enumerator.StopListenDevices()
	time.Sleep(1 * time.Second)
	s.SetStep(FirmwareFlashing)
	err := s.SelectedDevice.Flash(*s.FirmwarePath, func(message usb.FlashCallback) {
		if message.Type == usb.Log {
			s.Log("info", message.Message)
		}
		if message.Type == usb.Progress {
			uiEvent.Emit("flashProgress", &ProgressEvent{Current: message.Sent, Total: message.Total})
		}
	})
	if err != nil {
		s.Log("fatal", err.Error())
		s.SetStep(FatalError)
	} else {
		s.Log("info", "flash complete")
		s.SetStep(FlashComplete)
	}
	s.enumerator.ListenDevices()
}

func (s *State) HandleUSBConnectionEvent(connect bool, dev usb.Device) {
	if s.Devices == nil {
		s.Devices = make(map[int]usb.USBDevice)
	}

	bootloader := dev.GetBootloader()
	fingerprint := dev.GetFingerprint()
	firmwareFormat := dev.GetFile_format()
	model := dev.GetModel()
	name := dev.GetFriendly_name()
	portNumber := dev.GetPort_number()
	protocol := dev.GetProtocol()

	if connect {
		device := usb.USBDevice{FriendlyName: name, Model: model, Fingerprint: fingerprint, PortNumber: portNumber, Bootloader: bootloader, FirmwareFormat: firmwareFormat, Protocol: protocol, Handle: dev}
		s.Devices[fingerprint] = device
		uiEvent.Emit("deviceConnected", &DeviceConnectionEvent{Device: device})
		s.Log("info", "New device detected:")
		s.Log("info", fmt.Sprintf("'%s' | pointer: %d", name, fingerprint))
		s.Log("info", fmt.Sprintf("port: %d | bootloader: %t", portNumber, bootloader))

		// Trigger the flashing process if:
		// - The current step is KeyboardReset
		// - The connected device is on the same port number as the selected device
		// - The connected device is a bootloader
		if bootloader && s.SelectedDevice != nil && s.SelectedDevice.PortNumber == portNumber && s.Step == KeyboardReset {
			s.SelectedDevice = &device
			time.Sleep(1 * time.Second)
			s.SelectDevice(fingerprint)
			s.StartFlashing()
		}
	} else {
		uiEvent.Emit("deviceDisconnected", &DeviceDisconnectionEvent{Fingerprint: fingerprint})
	}
}

func (s *State) InstallUpdate() {

	var cmd *exec.Cmd
	switch runtime.GOOS {
	case "darwin":
		cmd = exec.Command("hdiutil", "attach", s.updatePath)
	case "windows":
		cmd = exec.Command(s.updatePath)
	}

	err := cmd.Run()
	if err != nil {
		s.Log("fatal", err.Error())
		return
	}

	wails.Quit(s.ctx)
}

/*
func (s *State) Open(fingerprint int) bool {
	dev := s.Devices[fingerprint]
	return dev.Open()
}
*/

func (s *State) Reset() {
	s.SelectedDevice = nil
	s.FirmwarePath = nil
	uiEvent.Emit("reset", nil)
	s.SetStep(Probing)
}

func (s *State) Quit() {
	s.Teardown()
	wails.Quit(s.ctx)
}

func (s *State) Teardown() {
	//usb.DeleteEnumerator(s.enumerator)
}
